pkg/runner/test_runner.go

```diff
@@ -6,282 +6,778 @@
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			return nil, fmt.Errorf("%s: result for nonexistent target %s %v", ss, tgt, res)
+		}
+		if ss.tgtDone[tgt] {
+			return nil, &cerrors.ErrTestStepReturnedDuplicateResult{
+				StepName: ss.sb.TestStepLabel,
+				Target:   tgt.ID,
+			}
+		}
+		ss.tgtDone[tgt] = true
+		// Begin is also allowed here because it may happen that we get a result before target runner updates phase.
+		if ts.CurStep != ss.stepIndex || (ts.CurPhase != targetStepPhaseBegin && ts.CurPhase != targetStepPhaseRun) {
+			return nil, &cerrors.ErrTestStepReturnedUnexpectedResult{
+				StepName: ss.sb.TestStepLabel,
+				Target:   tgt.ID,
+			}
+		}
+		if ts.resCh == nil {
+			// This should not happen, must be an internal error.
+			return nil, fmt.Errorf("%s: target runner %s is not there, dropping result on the floor", ss, ts)
+		}
+		ss.log.Debugf("%s: result for %s: %v", ss, ts, res)
+		return ts.resCh, nil
+	}()
+	if err != nil {
+		return err
 	}
+	select {
+	case resCh <- res:
+		break
+	case <-ctx.Done():
+		break
+	}
+	return nil
 }
 
-// NewTestRunnerWithTimeouts initializes and returns a new TestRunner object with
-// custom timeouts
-func NewTestRunnerWithTimeouts(timeouts TestRunnerTimeouts) TestRunner {
-	return TestRunner{timeouts: timeouts}
+func (tr *testRunner) safeCloseOutCh(ss *stepState) {
+	defer func() {
+		if r := recover(); r != nil {
+			tr.mu.Lock()
+			ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+			tr.mu.Unlock()
+		}
+	}()
+	close(ss.outCh)
 }
 
-// State is a structure that models the current state of the test runner
-type State struct {
-	completedSteps   map[string]error
-	completedRouting map[string]error
-	completedTargets map[*target.Target]error
+// safeCloseErrCh closes error channel safely, even if it has already been closed.
+func (tr *testRunner) safeCloseErrCh(ss *stepState) {
+	defer func() {
+		if r := recover(); r != nil {
+			tr.mu.Lock()
+			ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+			tr.mu.Unlock()
+		}
+	}()
+	close(ss.errCh)
 }
 
-// NewState initializes a State object.
-func NewState() *State {
-	r := State{}
-	r.completedSteps = make(map[string]error)
-	r.completedRouting = make(map[string]error)
-	r.completedTargets = make(map[*target.Target]error)
-	return &r
+// stepReader receives results from the step's output channel and forwards them to the appropriate target runners.
+func (tr *testRunner) stepReader(ctx statectx.Context, ss *stepState) {
+	ss.log.Debugf("%s: step reader active", ss)
+	var err error
+	outCh := ss.outCh
+loop:
+	for {
+		select {
+		case tgt, ok := <-outCh:
+			if !ok {
+				ss.log.Debugf("%s: out chan closed", ss)
+				// At this point we may still have an error to report,
+				// wait until error channel is emptied too.
+				outCh = make(chan *target.Target)
+				tr.safeCloseErrCh(ss)
+				continue loop
```

Comment:
`loop` is not required here

Reply:
it's not but i think it's more readable that way